coralproject/talk
# [CORL-2305]: Email domain moderation settings for new users

## What does this PR do?

These changes add the ability to configure email domain moderation settings for new accounts in the admin. They make it possible to add certain email domains that, if new users sign up with an email address that uses these domains, they can be configured to be either always pre-moderated or banned. This is for https://vmproduct.atlassian.net/browse/CORL-2305

## What changes to the GraphQL/Database Schema does this PR introduce?

To the schema, this adds `emailDomainModeration` to the settings to keep track of the email domain moderation settings for a tenant. For this, it also adds an `EmailDomain` type and `NEW_USER_MODERATION` enum for the `BAN` or `ALWAYS_PREMOD` options. It also adds `emailDomain` to the `Query` type for the configuration form routes. It also adds `createEmailDomain`, `updateEmailDomain`, and `deleteEmailDomain` mutations.

To the models, it adds creating, updating, and deleting email domains.

## Does this PR introduce any new environment variables or feature flags? 

no

## How do I test this PR?

To test this PR, you can go to the admin --> configure --> moderation tab. Scroll to the bottom of the page, to the new `Email domain` section. See that no table is shown before you have any email domains added, and that there are also no errors. Add a new email domain. See it appear in the table. Check that if you sign up a new user with that email address domain, their account is set to banned/always pre-moderated.

Update the email domain that you added, and see that it's updated as expected. Add/update more email domains. Delete an email domain, and see that it's deleted as expected. See that if you click `Cancel` while adding/updating domains, you are taken back to the `Email domain` configuration section. You should also be taken there after adding/updating an email domain.

Check that the validation works as expected. For the email domain, it should check that there is no `@` included at the beginning, and that there is at least one `.` included in the domain. It should also check that the domain exists, as it's required. You should also see that if you try to add or update to an email domain that is already configured, you will see an error on add/update.

See that if you sign up with an account with an email address domain that's not included in these email domains, everything still works as expected.

Check that you can view the Account History with ban/premod status there, and that it says the ban/premod status was set by the `System`, and there are no errors.
 
## How do we deploy this PR?

n/a
